Microsoft.Toolkit.Mvvm/ComponentModel/ObservableValidator.cs

```diff
@@ -0,0 +1,159 @@
+        /// </summary>
+        private int totalErrors;
+
+        /// <inheritdoc/>
+        public event EventHandler<DataErrorsChangedEventArgs>? ErrorsChanged;
+
+        /// <inheritdoc/>
+        public bool HasErrors => this.totalErrors > 0;
+
+        /// <inheritdoc/>
+        [Pure]
+        public IEnumerable GetErrors(string? propertyName)
+        {
+            // Entity-level errors when the target property is null or empty
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                return this.GetAllErrors();
+            }
+
+            // Property-level errors, if any
+            if (this.errors.TryGetValue(propertyName!, out List<ValidationResult> errors))
+            {
+                return errors;
+            }
+
+            // The INotifyDataErrorInfo.GetErrors method doesn't specify exactly what to
+            // return when the input property name is invalid, but given that the return
+            // type is marked as a non-nullable reference type, here we're returning an
+            // empty array to respect the contract. This also matches the behavior of
+            // this method whenever errors for a valid properties are retrieved.
+            return Array.Empty<ValidationResult>();
+        }
+
+        /// <summary>
+        /// Implements the logic for entity-level errors gathering for <see cref="GetErrors"/>.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerable"/> instance with all the errors in <see cref="errors"/>.</returns>
+        [Pure]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private IEnumerable GetAllErrors()
+        {
+            return this.errors.Values.SelectMany(errors => errors);
+        }
+
+        /// <summary>
+        /// Validates a property with a specified name and a given input value.
+        /// </summary>
+        /// <param name="value">The value to test for the specified property.</param>
+        /// <param name="propertyName">The name of the property to validate.</param>
+        /// <returns>Whether or not the validation was successful.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="propertyName"/> is <see langword="null"/>.</exception>
+        protected bool ValidateProperty(object? value, [CallerMemberName] string? propertyName = null)
+        {
+            if (propertyName is null)
+            {
+                ThrowArgumentNullExceptionForNullPropertyName();
+            }
+
+            bool errorsChanged = false;
+
+            // Clear the errors for the specified property, if any
+            if (this.errors.TryGetValue(propertyName!, out List<ValidationResult>? propertyErrors) &&
+                propertyErrors.Count > 0)
+            {
+                propertyErrors.Clear();
+
+                errorsChanged = true;
+            }
+
+            List<ValidationResult> results = new List<ValidationResult>();
```

Comment:
Could you reuse the propertyErrors list if it already exists, instead of doing an extra allocation here?

Reply:
Ah, good catch! 😄

Fixed it in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3429/commits/77ed6e8bfdd9e129d2dfb350cbe8e0f673be0c08, we're now always reusing the same `List<ValidationResult>` instance, and we're also not doing the `AddRange` call anymore either, so we should also get a small perf boost in the process 🚀